app/controllers/sitemap_controller.rb

```diff
@@ -0,0 +1,107 @@
+class SitemapController < ActionController::Metal
+
+  include AbstractController::Rendering
+  include ActionController::MimeResponds
+  include ActionController::DataStreaming
+  include ActionController::RackDelegation
+  include ActionController::Rescue
+  include ActionController::Head
+  include ActionController::Redirecting
+
+  def sitemap
+    community = request.env[:current_marketplace]
+
+    return unless can_show_sitemap?(community)
+
+    if APP_CONFIG.asset_host.present?
+      redirect_to ActionController::Base.helpers.asset_url(
+                    "/sitemap/generate.xml?sitemap_host=#{request.host}")
+    else
+      render_site_map(community)
+    end
+  end
+
+  def generate
+    community = CurrentMarketplaceResolver.resolve_from_host(
+      params[:sitemap_host], URLUtils.strip_port_from_host(APP_CONFIG.domain))
+
+    return unless can_show_sitemap?(community)
+
+    render_site_map(community)
+  end
+
+  private
+
+  def render_site_map(community)
+    default_host = community.full_domain(with_protocol: true)
+
+    sitemap = from_cache([community.id, max_sitemap_links, default_host]) do
+      adapter = SitemapGenerator::NeverWriteAdapter.new
+
+      open_listings = find_open_listings(community.id)
+
+      SitemapGenerator::Sitemap.create(
+            default_host: default_host,
+            verbose: false,
+            adapter: adapter) do
+        open_listings.each do |l|
+          add listing_path(id: l[:id]), lastmod: l[:lastmod]
+        end
+      end
+
+      adapter.data
+    end
+
+    send_data(sitemap, filename: "sitemap.xml")
```

Comment:
We cache the content for 24 hours to memcached which is great. In addition to this we should also set the cache-control headers correctly. Right now we send back cache-control: private which means that no client is allowed to cache the response. The minimum we should do is to set expiry to 24h. This is easy to do by including ActionController::ConditionalGet and using expire_in method. Better still would be to handle conditional gets on controller level (now there's rack middleware but since we're not sending a string body I don't think it kicks in. This is a bit more work so I'll leave it up to you to consider if it's worth doing. LandingPageController can be checked for reference.